x/staking/keeper/msg_server.go

```diff
@@ -324,13 +373,47 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
 	if err != nil {
 		return nil, err
 	}
+
+	tokens := msg.Amount.Amount
 	shares, err := k.ValidateUnbondAmount(
-		ctx, delegatorAddress, addr, msg.Amount.Amount,
+		ctx, delegatorAddress, addr, tokens,
 	)
 	if err != nil {
 		return nil, err
 	}
 
+	validator, found := k.GetValidator(ctx, addr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	delegation, found := k.GetDelegation(ctx, delegatorAddress, addr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorAddress,
+		)
+	}
+
+	// if this is a validator self-bond, the new liquid delegation cannot fall below the self-bond * bond factor
+	if delegation.ValidatorBond {
+		if err := k.SafelyDecreaseValidatorBond(ctx, validator, shares); err != nil {
+			return nil, err
+		}
+	}
+
```

Comment:
would it make more sense to pass a pointer instead of a copy?

Reply:
Probably it's better to not pass the validator struct at all but rather call GetValidator in the function. Like this the function is much safer to use.

Reply:
I think the options are to either call GetValidator _after_ the function call or to pass a pointer. I don't think calling GetValidator in the function solves anything here cause the outer function would still have a stale struct.

Imo, using a pointer makes more sense, otherwise the burden is on the developer to remember to call Get again after every call to SafelyIncrease/Decrease

Reply:
@alexanderbez what's the way SDK is dealing with such methods?

Reply:
I see that a common used pattern is: `validator = fn(validator)`.

Reply:
updated [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/3024452ffacf16158819e0ca92842e05bd38eb50)